src/command_modules/azure-cli-extension/azure/cli/command_modules/extension/custom.py

```diff
@@ -0,0 +1,176 @@
+        min_max_msg_fmt = "The extension is not compatible with this version of the CLI.\n" \
+                          "You have CLI version {} and this extension " \
+                          "requires ".format(cli_version)
+        if min_required and max_required:
+            min_max_msg_fmt += 'a min of {} and max of {}.'.format(min_required, max_required)
+        elif min_required:
+            min_max_msg_fmt += 'a min of {}.'.format(min_required)
+        elif max_required:
+            min_max_msg_fmt += 'a max of {}.'.format(max_required)
+        raise CLIError(min_max_msg_fmt)
+
+
+def _validate_whl_extension(ext_file):
+    tmp_dir = tempfile.mkdtemp()
+    zip_ref = zipfile.ZipFile(ext_file, 'r')
+    zip_ref.extractall(tmp_dir)
+    zip_ref.close()
+    azext_metadata = WheelExtension.get_azext_metadata(tmp_dir)
+    _validate_whl_cli_compat(azext_metadata)
+
+
+def _add_whl_ext(source):  # pylint: disable=too-many-statements
+    import pip
+    url_parse_result = urlparse(source)
+    is_url = (url_parse_result.scheme == 'http' or url_parse_result.scheme == 'https')
+    logger.debug('Extension source is url? %s', is_url)
+    whl_filename = os.path.basename(url_parse_result.path) if is_url else os.path.basename(source)
+    parsed_filename = WHEEL_INFO_RE(whl_filename)
+    extension_name = parsed_filename.groupdict().get('name') if parsed_filename else None
+    if not extension_name:
+        raise CLIError('Unable to determine extension name from {}. Is the file name correct?'.format(source))
+    if extension_exists(extension_name):
+        raise CLIError('The extension {} already exists.'.format(extension_name))
+    ext_file = None
+    if is_url:
+        # Download from URL
+        tmp_dir = tempfile.mkdtemp()
+        ext_file = os.path.join(tmp_dir, whl_filename)
+        logger.debug('Downloading %s to %s', source, ext_file)
+        try:
+            _whl_download_from_url(url_parse_result, ext_file)
+        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError) as err:
+            raise CLIError('Please ensure you have network connection. Error detail: {}'.format(str(err)))
+        logger.debug('Downloaded to %s', ext_file)
+    else:
+        # Get file path
+        ext_file = os.path.realpath(os.path.expanduser(source))
+        if not os.path.isfile(ext_file):
+            raise CLIError("File {} not found.".format(source))
+    # Validate the extension
+    logger.debug('Validating the extension {}'.format(ext_file))
+    try:
+        _validate_whl_extension(ext_file)
+    except AssertionError:
+        logger.debug(traceback.format_exc())
+        raise CLIError('The extension is invalid. Use --debug for more information.')
+    except CLIError as e:
+        raise e
+    except Exception:  # pylint:disable=broad-except
+        logger.debug(traceback.format_exc())
+        raise CLIError('Error validating extension. Use --debug for more information.')
```

Comment:
Why don't we just let the except been thrown as is? What's the benefit of converting it to `CLIError` and swallow the original exception?